_SPIKE_.md

```diff
@@ -0,0 +1,527 @@
+```ts
+/**
+ * Description of capabilities offered by the connector backing the given
+ * datasource.
+ */
+export interface ConnectorCapabilities {
+  /**
+   * Maximum number of items allowed for `inq` operators.
+   * This value is used to split queries when resolving related models
+   * for a large number of source instances.
+   */
+  inqLimit?: number;
+}
+```
+
+To preserve backwards compatibility with existing repository implementation, we
+cannot add `ConnectorCapabilities` directly to the `Repository` class. We need
+to introduce a new interface instead that Repositories can (or may not)
+implement.
+
+```ts
+export interface WithCapabilities {
+  capabilities: ConnectorCapabilities;
+}
+```
+
+### MongoDB and `ObjectID` type
+
+MongoDB is tricky.
+
+- It uses a custom `ObjectID` type for primary keys.
+- `ObjectID` is represented as a `string` when converted to JSON
+- In queries, string values must be case to ObjectID, otherwise they are not
+  considered as the same value: `'some-id' !== ObjectID('some-id')`.
+
+As a result, both PK and FK properties must use `ObjectID` as the type, and
+coercion must be applied where necessary.
+
+Ideally, I'd like LB4 to define MongoDB PK and FKs as follows:
+
+- `{type: 'string', mongodb: {dataType: 'ObjectID'}}`
+
+Even better, `dataType: 'ObjectID'` should be automatically applied by the
+connector for PK and FKs referencing ObjectID PKs.
+
+For example:
+
+```ts
+@model()
+class Product {
+  @property({
+    type: 'string',
+    generated: true,
+    // ^^ implies dataType: 'ObjectID'
+  })
+  id: string;
+
+  @property({
+    type: 'string',
+    references: {
+      model: () => Category,
+      property: 'id',
+    },
+    // ^^ implies dataType: 'ObjectID' when Category is attached to MongoDB
+  })
+  categoryId: string;
+}
+```
+
+For v1, I suppose we can ask developers to provide dataType manually.
+
+```ts
+@model()
+class Product {
+  @property({
+    type: 'string',
+    generated: true,
+    mongodb: {dataType: 'ObjectID'},
+  })
+  id: string;
+
+  @property({
+    type: 'string',
+    mongodb: {dataType: 'ObjectID'},
+  })
+  categoryId: string;
+}
+```
+
+With this setup in place, `id` and `categoryId` properties should be always
+returned as strings from DAO and connector methods.
+
+This is tricky to achieve for the PK property, because juggler overrides
+user-supplied type with connector's default type when the PK is generated by the
+database. See
+[`DataSource.prototype.setupDataAccess()`](https://github.com/strongloop/loopback-datasource-juggler/blob/0c2bb81dace3592ecde8b9eccbd70d589da44d7d/lib/datasource.js#L713-L719)
+
+Can we rework MongoDB connector to hide ObjectID complexity as an internal
+implementation detail and always use string values in public APIs? Accept ids as
+strings and internally convert them to ObjectID. Convert values returned by the
+database from ObjectID to strings.
+
+Downside: this is not going to work for free-form properties that don't have any
```

Comment:
In that case, the type should be interpreted as "whatever the DB driver returns". So, I don't really see it as a downside.

Reply:
Well, from the point of view of a REST API client, "whatever the DB driver returns" is eventually converted to "JSON representation of whatever the DB driver returns", which is a string for ObjectID.

If we stay only in JS/TS land, then I agree with you. The DB driver returns the same type that was used to supply the (initial) value of the free-form property, thus all should be good.